docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,211 @@
+
+- Graph space management
+
+      Metadata service stores the metadata of all spaces in the cluster and tracks changes that take place in these spaces, like adding, dropping space, modifying graph space configuration (Raft copies).
+
+-   Schema management
+
+      Nebula Graph is a strong typed database.
+
+      - Types of tag and edge properties are recorded by meta service. Supported data types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The meta service is stateful, and just like the storage service, it persists data to a key-value store.
+
+
+![image](https://user-images.githubusercontent.com/42762957/61120413-8cdd2700-a4cf-11e9-8846-14b5d8bd6693.png)
+
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the query engine is nebula-graphd, which consists
+of nodes that are stateless, equally privileged and isolated from each other.
+
+Query engine accepts the message from the client and generates the query plan after lexical analysis (Lexer), parsing (Parser) and query optimization. Then the execution plan will be passed to the execution engine. The query execution engine takes the query plans and interacts with meta server and the storage engine to retrieve the schema and data.
+
+The main optimizations of the query engine are:
+
+- Asynchronous and parallel execution
+
+      I/O operations and network transmission are time-consuming. Thus asynchronous and parallel operations are adopted in Nebula query engine to guarantee quality of service (QoS).Also, a separate resource pool is set for each query to avoid the long tail effect of those time-consuming queries.
+
+- Pushing down computation
+
+      Too much data passing back form storage layer to computation layer occupies  too much bandwidth, to avoid that, condition filtering such as 'WHERE' clause is pushed down to the storage layer along with the conditions.
+
+- Query optimizer
+
+      Query optimizer for SQL has experienced a long-term development, but not for graph database. Nebula Graph has made an effort in the optimizations of graph query, including cache executing plan and executing context-free queries in parallel.
+
+
+![image](https://user-images.githubusercontent.com/42762957/61119795-26a3d480-a4ce-11e9-97e9-102bf14e72d8.png)
+
+
+
+
+### API and SDK
+
+Nebula Graph provides SDKs in C++, Java, and Golang. Nebula Graph uses RPC to communicate among servers, and the communication protocol is Facebook-Thrift. Users can also use Nebula Graph on Linux. Nebula Graph's web SDK is in progress and will be released soon.
```

Comment:
Nebula Graph provides SDKs in C++, Java, and Golang. Nebula Graph uses fbthrift as the RPC framework to communicate among servers. Nebula Graph's web SDK is in progress and will be released soon.